### PR TITLE
graph: use int64 for node retrieval

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

This is a major breaking change to the graph packages.

This was not done mechanically as I don't think `eg` is able to handle some of the subtlety here. Because of this it needs to be read reasonably carefully in some places, particularly in graph/community where I introduced a bug that was only found because of tests (I would have hated to do this in a dynamically typed language).

Updates #31.

### CODE REVIEW

**Comment by DryHumour** [comment]
Will the gonum API version number be bumped for this?

**Comment by kortschak** [comment]
We don't have a version number yet. The v1 in the import path is aspirational - and yes, we're running late. We are aiming for an API stabilisation, but there is I think at least one more large scale breaking change to come in graph. I'm sorry for the trouble this causes, but the design process has been difficult, and something that I have had to fit in around my paying work with help from only two other people. Please bear that in mind.

**Comment by kortschak** [comment]
Also, it is likely that for most uses, the change here will make little difference to the client code (unless the client is implementing `graph.Graph` or `graph.Multigraph` for themself.

**Comment by DryHumour** [comment]
No worries, just wanted to check what the plans were.  Thanks for the information!  (And for your hard work.)

**Comment by kortschak** [comment]
@vladimir-ch @mewmew Do we want to add `Remove` to the list to methods that take an `id int64` as well? Any others?

**Comment by mewmew** [comment]
@kortschak I'm going up to the northen Sweden to visit my family, so won't have the chance to look at this in the time being. Just to let you know. I feel happy you are pushing in this direction obviously, as it is helping resolve a long standing issue.

Cheers,
/u

**Comment by kortschak** [comment]
Var i norra Sverige. Jag var i Holmsund för många år sedan. (Jag läser svenska igen nu).

**Comment by mewmew** [comment]
> Var i norra Sverige? Jag var i Holmsund för många år sedan. (Jag läser svenska igen nu).

Kul att höra! Vi åker upp till Östersund (så snarare mittpunkten av Sverige, men norr i förhållande till Stockholm). Lär du dig svenska nu genom att gå någon kurs? Eller mest på fritiden genom att läsa?

**Comment by kortschak** [comment]
På fritiden genom att läsa.

File: graph/simple/weighted_directed_test.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ var (
 func TestWeightedEdgeOvercounting(t *testing.T) {
 	g := generateDummyGraph()
 
-	if neigh := g.From(Node(Node(2))); len(neigh) != 2 {
```
Comment (mewmew) [inline graph/simple/weighted_directed_test.go]: Out of curiosity, what was it `Node(Node(foo))` before and not just `Node(foo)`?

File: graph/multigraph.go [hunk 2]
```diff
@@ -19,78 +19,83 @@ type WeightedLine interface {
...
-	HasEdgeBetween(x, y Node) bool
+	// nodes with IDs xid and yid without considering direction.
+	HasEdgeBetween(xid, yid int64) bool
 
-	// Lines returns the lines from u to v if any such lines
-	// exist and nil otherwise. The node v must be directly
-	// reachable from u as defined by the From method.
-	Lines(u, v Node) []Line
+	// Lines returns the edge from u to v, with IDs uid and vid,
+	// if such an edge exists and nil otherwise. The node v
```
Comment (mewmew) [inline graph/multigraph.go]: Was it intentional to change from line to edge in the description?

File: graph/simple/weighted_directed_test.go (hunk 1 above)
Comment (kortschak) [reply graph/simple/weighted_directed_test.go]: ‽

Those tests need cleaning up.

File: graph/multigraph.go (hunk 2 above)
Comment (kortschak) [reply graph/multigraph.go]: No, it wasn't. Fixed, thanks.gr

**Comment by kortschak** [comment]
I don't trust the compiler to do those optimisations properly at the moment. I guess that's something we can follow up though. Do you want to file a ponies issue about that?

**Comment by mewmew** [comment]
> I don't trust the compiler to do those optimisations properly at the moment. I guess that's something we can follow up though. Do you want to file a ponies issue about that?

Sure.

**Comment by mewmew** [comment]
Done https://github.com/gonum/gonum/issues/437

File: graph/multigraph.go [hunk 3]
```diff
@@ -19,78 +19,83 @@ type WeightedLine interface {
...
+	// within the multigraph.
+	Has(id int64) bool
 
 	// Nodes returns all the nodes in the multigraph.
 	Nodes() []Node
 
 	// From returns all nodes that can be reached directly
-	// from the given node.
-	From(Node) []Node
+	// from the node with the given ID
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/multigraph.go]: Fixed

**Comment by kortschak** [comment]
PTAL

**Comment by mewmew** [comment]
Great work @kortschak!

**Pull request merged**